File: js/bff.js
/**
 * BFFInterpreter - Brainfuck Friends interpreter
 *
 * Implements the BFF instruction set from the Turing Soup paper.
 * Uses three heads on a single tape:
 * - IP (instruction pointer): current execution position
 * - head0 (read head): for reading data
 * - head1 (write head): for writing data
 *
 * This is a pure interpreter with no visualization logic.
 */

// BFF instruction byte values
export const BFF_INSTRUCTIONS = {
  '<': 0x3C,  // head0--
  '>': 0x3E,  // head0++
  '{': 0x7B,  // head1--
  '}': 0x7D,  // head1++
  '-': 0x2D,  // tape[head0]--
  '+': 0x2B,  // tape[head0]++
  '.': 0x2E,  // tape[head1] = tape[head0]
  ',': 0x2C,  // tape[head0] = tape[head1]
  '[': 0x5B,  // if tape[head0]==0, jump forward to matching ]
  ']': 0x5D,  // if tape[head0]!=0, jump back to matching [
};

// Reverse lookup: byte value to instruction character
export const BYTE_TO_INSTRUCTION = {};
for (const [char, byte] of Object.entries(BFF_INSTRUCTIONS)) {
  BYTE_TO_INSTRUCTION[byte] = char;
}

// Instruction categories for visualization
export const INSTRUCTION_CATEGORIES = {
  'head_movement': ['<', '>', '{', '}'],
  'arithmetic': ['+', '-'],
  'copy': ['.', ','],
  'control': ['[', ']'],
};

/**
 * Check if a byte value is a BFF instruction
 * @param {number} byte - Byte value 0-255
 * @returns {string|null} Instruction character or null if no-op
 */
export function byteToInstruction(byte) {
  return BYTE_TO_INSTRUCTION[byte] || null;
}

/**
 * Get the category of an instruction
 * @param {string} instruction - Single character instruction
 * @returns {string|null} Category name or null
 */
export function getInstructionCategory(instruction) {
  for (const [category, instructions] of Object.entries(INSTRUCTION_CATEGORIES)) {
    if (instructions.includes(instruction)) {
      return category;
    }
  }
  return null;
}

// v12 - MAX_STEPS = 8192 to match paper
export class BFFInterpreter {
  static MAX_STEPS = 8192; // 2^13 from paper

  /**
   * Create a new BFF interpreter
   * @param {Tape} tape - The tape to execute on
   */
  constructor(tape) {
    this.tape = tape;
    this.reset();
  }

  /**
   * Reset all heads to initial positions
   */
  reset() {
    this.ip = 0;      // Instruction pointer
    this.head0 = 0;   // Read head
    this.head1 = 0;   // Write head
    this.stepCount = 0;
    this.writeCount = 0;  // Track tape modifications
    this.loopJumps = 0;   // Track backward jumps from ]
    this.halted = false;
    this.haltReason = null;
  }

  /**
   * Get current interpreter state
   * @returns {Object} State object with all head positions and status
   */
  getState() {
    return {
      ip: this.ip,
      head0: this.head0,
      head1: this.head1,
      stepCount: this.stepCount,
      halted: this.halted,
      haltReason: this.haltReason,
      currentInstruction: byteToInstruction(this.tape.get(this.ip)),
      currentByte: this.tape.get(this.ip),
    };
  }

  /**
   * Execute one instruction
   * @returns {Object} State after execution, including what changed
   */
  step() {
    if (this.halted) {
      return { ...this.getState(), changed: null };
    }

    // Check step limit
    if (this.stepCount >= BFFInterpreter.MAX_STEPS) {
      this.halted = true;
      this.haltReason = 'max_steps';
      return { ...this.getState(), changed: null };
    }

    const byte = this.tape.get(this.ip);
    const instruction = byteToInstruction(byte);
    const changed = { type: null, index: null, oldValue: null, newValue: null };

    this.stepCount++;

    if (instruction === null) {
      // No-op: just advance IP
      this.ip++;
      if (this.ip >= this.tape.size) {
        this.halted = true;
        this.haltReason = 'end_of_tape';
      }
      return { ...this.getState(), changed };
    }

    switch (instruction) {
      case '<': // head0--
        this.head0 = this._wrapHead(this.head0 - 1);
        break;

      case '>': // head0++
        this.head0 = this._wrapHead(this.head0 + 1);
        break;

      case '{': // head1--
        this.head1 = this._wrapHead(this.head1 - 1);
        break;

      case '}': // head1++
        this.head1 = this._wrapHead(this.head1 + 1);
        break;

      case '-': // tape[head0]--
        changed.type = 'decrement';
        changed.index = this.head0;
        changed.oldValue = this.tape.get(this.head0);
        this.tape.decrement(this.head0);
        changed.newValue = this.tape.get(this.head0);
        this.writeCount++;
        break;

      case '+': // tape[head0]++
        changed.type = 'increment';
        changed.index = this.head0;
        changed.oldValue = this.tape.get(this.head0);
        this.tape.increment(this.head0);
        changed.newValue = this.tape.get(this.head0);
        this.writeCount++;
        break;

      case '.': // tape[head1] = tape[head0]
        changed.type = 'copy_to_head1';
        changed.index = this.head1;
        changed.oldValue = this.tape.get(this.head1);
        this.tape.set(this.head1, this.tape.get(this.head0));
        changed.newValue = this.tape.get(this.head1);
        this.writeCount++;
        break;

      case ',': // tape[head0] = tape[head1]
        changed.type = 'copy_to_head0';
        changed.index = this.head0;
        changed.oldValue = this.tape.get(this.head0);
        this.tape.set(this.head0, this.tape.get(this.head1));
        changed.newValue = this.tape.get(this.head0);
        this.writeCount++;
        break;

      case '[': // if tape[head0]==0, jump forward to matching ]
        if (this.tape.get(this.head0) === 0) {
          const target = this._findMatchingBracket(this.ip, 1);
          if (target === -1) {
            this.halted = true;
            this.haltReason = 'unmatched_bracket';
            return { ...this.getState(), changed };
          }
          this.ip = target;
        }
        break;

      case ']': // if tape[head0]!=0, jump back to matching [
        if (this.tape.get(this.head0) !== 0) {
          const target = this._findMatchingBracket(this.ip, -1);
          if (target === -1) {
            this.halted = true;
            this.haltReason = 'unmatched_bracket';
            return { ...this.getState(), changed };
          }
          this.ip = target;
          this.loopJumps++;
        }
        break;
    }

    // Advance IP (unless halted)
    if (!this.halted) {
      this.ip++;
      if (this.ip >= this.tape.size) {
        this.halted = true;
        this.haltReason = 'end_of_tape';
      }
    }

    return { ...this.getState(), changed };
  }

  /**
   * Find matching bracket (does not cross tape boundaries)
   * @param {number} start - Starting position
   * @param {number} direction - 1 for forward, -1 for backward
   * @returns {number} Position of matching bracket, or -1 if not found
   * @private
   */
  _findMatchingBracket(start, direction) {
    let depth = 1;
    let pos = start;
    const openBracket = BFF_INSTRUCTIONS['['];
    const closeBracket = BFF_INSTRUCTIONS[']'];

    // Search without wrapping - stop at tape boundaries
    while (true) {
      pos += direction;

      // Stop at tape boundaries
      if (pos < 0 || pos >= this.tape.size) {
        return -1; // Unmatched - hit boundary
      }

      const byte = this.tape.get(pos);

      if (direction === 1) {
        // Looking for ]
        if (byte === openBracket) depth++;
        else if (byte === closeBracket) depth--;
      } else {
        // Looking for [
        if (byte === closeBracket) depth++;
        else if (byte === openBracket) depth--;
      }

      if (depth === 0) {
        return pos;
      }
    }
  }

  /**
   * Wrap IP to valid range
   * @private
   */
  _wrapIP(index) {
    return ((index % this.tape.size) + this.tape.size) % this.tape.size;
  }

  /**
   * Wrap head position to valid range
   * @private
   */
  _wrapHead(index) {
    return ((index % this.tape.size) + this.tape.size) % this.tape.size;
  }

  /**
   * Run until halted or max steps
   * @param {function} onStep - Optional callback after each step
   * @returns {Object} Final state
   */
  run(onStep = null) {
    while (!this.halted) {
      const state = this.step();
      if (onStep) onStep(state);
    }
    return this.getState();
  }
}

File: wasm/src/bff.rs
//! BFF (Brainfuck Friends) Interpreter
//!
//! Implements the BFF instruction set from the Turing Soup paper.
//! Uses three heads on a single tape:
//! - IP (instruction pointer): current execution position
//! - head0 (read head): for reading data
//! - head1 (write head): for writing data

/// Maximum steps before halting (2^13, matching paper)
pub const MAX_STEPS: u32 = 8192;

/// BFF instruction byte values
pub mod instructions {
    pub const HEAD0_DEC: u8 = b'<';   // 0x3C - head0--
    pub const HEAD0_INC: u8 = b'>';   // 0x3E - head0++
    pub const HEAD1_DEC: u8 = b'{';   // 0x7B - head1--
    pub const HEAD1_INC: u8 = b'}';   // 0x7D - head1++
    pub const DECREMENT: u8 = b'-';   // 0x2D - tape[head0]--
    pub const INCREMENT: u8 = b'+';   // 0x2B - tape[head0]++
    pub const COPY_TO_H1: u8 = b'.';  // 0x2E - tape[head1] = tape[head0]
    pub const COPY_TO_H0: u8 = b',';  // 0x2C - tape[head0] = tape[head1]
    pub const LOOP_START: u8 = b'[';  // 0x5B - if tape[head0]==0, jump to ]
    pub const LOOP_END: u8 = b']';    // 0x5D - if tape[head0]!=0, jump to [
}

/// Halt reasons
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[repr(u8)]
pub enum HaltReason {
    EndOfTape = 0,
    MaxSteps = 1,
    UnmatchedBracket = 2,
    NoInstructions = 3,
}

/// Result of executing a tape
#[derive(Clone, Copy, Debug)]
pub struct ExecutionResult {
    pub steps: u32,
    pub head0_count: u32,  // < > (head0 movements)
    pub head1_count: u32,  // { } (head1 movements)
    pub math_count: u32,   // + - (increment/decrement)
    pub copy_count: u32,   // . , (copy operations)
    pub loop_count: u32,   // [ ] (loop operations)
    pub halt_reason: HaltReason,
}

/// Check if a byte is a BFF instruction
#[inline]
pub fn is_instruction(byte: u8) -> bool {
    matches!(
        byte,
        instructions::HEAD0_DEC
            | instructions::HEAD0_INC
            | instructions::HEAD1_DEC
            | instructions::HEAD1_INC
            | instructions::DECREMENT
            | instructions::INCREMENT
            | instructions::COPY_TO_H1
            | instructions::COPY_TO_H0
            | instructions::LOOP_START
            | instructions::LOOP_END
    )
}

/// Check if tape contains any BFF instructions
pub fn has_instructions(tape: &[u8]) -> bool {
    tape.iter().any(|&b| is_instruction(b))
}

/// Find matching bracket (does not cross tape boundaries)
/// Returns None if unmatched
#[inline]
fn find_matching_bracket(tape: &[u8], start: usize, direction: i32) -> Option<usize> {
    let mut depth = 1i32;
    let mut pos = start as i32;
    let size = tape.len() as i32;

    loop {
        pos += direction;

        // Stop at tape boundaries
        if pos < 0 || pos >= size {
            return None; // Unmatched - hit boundary
        }

        let byte = tape[pos as usize];

        if direction > 0 {
            // Looking for ]
            if byte == instructions::LOOP_START {
                depth += 1;
            } else if byte == instructions::LOOP_END {
                depth -= 1;
            }
        } else {
            // Looking for [
            if byte == instructions::LOOP_END {
                depth += 1;
            } else if byte == instructions::LOOP_START {
                depth -= 1;
            }
        }

        if depth == 0 {
            return Some(pos as usize);
        }
    }
}

/// Execute BFF program on tape
/// Modifies tape in place and returns execution statistics
/// head1_start: initial position of head1 (defaults to midpoint for paired execution)
pub fn execute(tape: &mut [u8]) -> ExecutionResult {
    execute_with_head1(tape, tape.len() / 2)
}

/// Execute BFF program - head0 starts at 0, head1 starts at specified position
/// Both heads can move freely across entire tape (toroidal wrapping)
pub fn execute_with_head1(tape: &mut [u8], head1_start: usize) -> ExecutionResult {
    execute_with_params(tape, head1_start, MAX_STEPS)
}

/// Execute BFF program with configurable head1 start and max steps
pub fn execute_with_params(tape: &mut [u8], head1_start: usize, max_steps: u32) -> ExecutionResult {
    let size = tape.len();

    // Early abort if no instructions
    if !has_instructions(tape) {
        return ExecutionResult {
            steps: 0,
            head0_count: 0,
            head1_count: 0,
            math_count: 0,
            copy_count: 0,
            loop_count: 0,
            halt_reason: HaltReason::NoInstructions,
        };
    }

    let mut ip: usize = 0;
    let mut head0: i32 = 0;  // Starts at beginning (left tape)
    let mut head1: i32 = head1_start as i32;  // Starts at specified position
    let mut steps: u32 = 0;
    let mut head0_count: u32 = 0;
    let mut head1_count: u32 = 0;
    let mut math_count: u32 = 0;
    let mut copy_count: u32 = 0;
    let mut loop_count: u32 = 0;

    // Wrap any head position to valid tape range
    let wrap = |h: i32| -> usize {
        ((h % size as i32) + size as i32) as usize % size
    };

    while steps < max_steps && ip < size {
        steps += 1;
        let byte = tape[ip];

        match byte {
            instructions::HEAD0_DEC => {
                head0 -= 1;
                head0 = wrap(head0) as i32;
                head0_count += 1;
            }
            instructions::HEAD0_INC => {
                head0 += 1;
                head0 = wrap(head0) as i32;
                head0_count += 1;
            }
            instructions::HEAD1_DEC => {
                head1 -= 1;
                head1 = wrap(head1) as i32;
                head1_count += 1;
            }
            instructions::HEAD1_INC => {
                head1 += 1;
                head1 = wrap(head1) as i32;
                head1_count += 1;
            }
            instructions::DECREMENT => {
                let idx = wrap(head0);
                tape[idx] = tape[idx].wrapping_sub(1);
                math_count += 1;
            }
            instructions::INCREMENT => {
                let idx = wrap(head0);
                tape[idx] = tape[idx].wrapping_add(1);
                math_count += 1;
            }
            instructions::COPY_TO_H1 => {
                let src = wrap(head0);
                let dst = wrap(head1);
                tape[dst] = tape[src];
                copy_count += 1;
            }
            instructions::COPY_TO_H0 => {
                let src = wrap(head1);
                let dst = wrap(head0);
                tape[dst] = tape[src];
                copy_count += 1;
            }
            instructions::LOOP_START => {
                let idx = wrap(head0);
                if tape[idx] == 0 {
                    match find_matching_bracket(tape, ip, 1) {
                        Some(target) => ip = target,
                        None => {
                            return ExecutionResult {
                                steps,
                                head0_count,
                                head1_count,
                                math_count,
                                copy_count,
                                loop_count,
                                halt_reason: HaltReason::UnmatchedBracket,
                            };
                        }
                    }
                }
            }
            instructions::LOOP_END => {
                let idx = wrap(head0);
                loop_count += 1;
                if tape[idx] != 0 {
                    match find_matching_bracket(tape, ip, -1) {
                        Some(target) => {
                            ip = target;
                        }
                        None => {
                            return ExecutionResult {
                                steps,
                                head0_count,
                                head1_count,
                                math_count,
                                copy_count,
                                loop_count,
                                halt_reason: HaltReason::UnmatchedBracket,
                            };
                        }
                    }
                }
            }
            _ => {
                // No-op: just advance IP
            }
        }

        ip += 1;
    }

    let halt_reason = if steps >= max_steps {
        HaltReason::MaxSteps
    } else {
        HaltReason::EndOfTape
    };

    ExecutionResult {
        steps,
        head0_count,
        head1_count,
        math_count,
        copy_count,
        loop_count,
        halt_reason,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_has_instructions() {
        assert!(has_instructions(b"abc+def"));
        assert!(has_instructions(b"["));
        assert!(!has_instructions(b"abcdef"));
        assert!(!has_instructions(b""));
    }

    #[test]
    fn test_simple_increment() {
        // head0 starts at 0, so '+' increments tape[0] from 0x2B to 0x2C
        let mut tape = vec![b'+', 0, 0, 0];
        let result = execute(&mut tape);
        assert_eq!(tape[0], 0x2C); // '+' (0x2B) becomes ',' (0x2C)
        assert_eq!(result.math_count, 1);
    }

    #[test]
    fn test_no_instructions() {
        let mut tape = vec![0, 1, 2, 3];
        let result = execute(&mut tape);
        assert_eq!(result.halt_reason, HaltReason::NoInstructions);
        assert_eq!(result.steps, 0);
    }
}
